subprojects/core/src/main/java/org/gradle/internal/service/scopes/TaskExecutionServices.java

```diff
@@ -139,9 +147,18 @@ TaskExecuter createTaskExecuter(TaskArtifactStateRepository repository, TaskOutp
         );
     }
 
-    private static TaskExecuter createSkipCachedExecuterIfNecessary(StartParameter startParameter, BuildCacheConfigurationInternal buildCacheConfiguration, TaskOutputPacker packer, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskOutputOriginFactory taskOutputOriginFactory, TaskExecuter delegate) {
+    private static TaskExecuter createSkipCachedExecuterIfNecessary(StartParameter startParameter, BuildCacheConfigurationInternal buildCacheConfiguration, TaskOutputPacker packer, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskOutputOriginFactory taskOutputOriginFactory, ListenerManager listenerManager, TaskExecuter delegate) {
         if (startParameter.isTaskOutputCacheEnabled()) {
-            return new SkipCachedTaskExecuter(taskOutputOriginFactory, buildCacheConfiguration, packer, taskOutputsGenerationListener, delegate);
+            TaskCachingReasonsListener cachingReasonsListener = listenerManager.getBroadcaster(TaskCachingReasonsListener.class);
```

Comment:
I agree the logic would be better if it wasn't in the service registry. How about moving the listener to its own service instead, and inject it as such? Adding another executer just for this seems to be the wrong abstraction level to handle this.